rle-bitpacked.md

```diff
@@ -0,0 +1,120 @@
+and the remaining bytes constitute the body.
+
+The first bit of the last byte of the header denotes whether the run is bitpacked-encoded
+or RLE-encoded. `h1 & 1 == 1` indicates a bitpacked-encoded run, `h1 & 1 != 1` a RLE-encoded run.
+
+### Decoding RLE-encoded runs
+
+Given a header `h1` and the stream of bytes past the header, the number of repetitions
+is given by `repetitions = h1 >> 1`. The number of bytes in the body, `body_length`,
+MUST be the minimum number of bytes that can hold `bit_width`, `body_length = ceil(bit_width, 8)`.
+The body MUST represent the repeated value in little endian for byte types (e.g. `int32`) and
+[LSB](https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit) for boolean types.
+
+### Decoding bitpacked-encoded runs
+
+Given a header `h1` and the stream of bytes past the header, the number of bytes
+in the body, `body_length`, is equal to `body_length = h1 >> 1`.
+The body represents bitpacked-encoded values with exactly `8 / bit_width` items.
+
+Note that for `bit_width = 1`, this encoding corresponds exactly to LSB.
+Thus, in-memory formats that store booleans or validities in `LSB` (e.g. Apache Arrow),
+the bodies can be memcopied _as is_ to a LSB buffer.
+
+### Example
+
+Let's now consider the the following byte stream and a `bit_width = 1`,
+
+```
+[
+    5, 0, 0, 0,
+    0b00000101, 0b11101011, 0b00000010,
+    0b00010000, 0b00000001,
+    0b00000101,
+    0b00000101,
+]
+```
+
+which could appear in a definition level of a non-nested Parquet type.
+
+We use the first 4 bytes to identify relevant length, within the stream,
+of the encoded bytes. This corresponds to `[5, 0, 0, 0]`, which, in little
+endian, corresponds to `5u32`. We can thus assume that all runs of this encoded
+stream are encoded in 
+
+```
+[
+    0b00000101, 0b11101011, 0b00000010,
+    0b00010000, 0b00000001,
+]
+```
+
+The ULEB128-decoding of this stream yields `h1 = 5i32` and length 1, i.e. the
+first byte is enough to represent the header of the first run. Because `5 & 1 == 1`,
+we conclude that the first run is bitpacked-encoded. `body_length = 5i32 >> 1 = 2`,
+which means that the next two bytes form the body. Bit-unpacking these two
+bytes yields (in `i32`), 
+
+```
+[
+    1, 1, 0, 1, 0, 1, 1, 1, 
+    0, 1, 0, 0, 0, 0, 0, 0
+]
+```
+
+(to understand why this is the case, read `0b11101011, 0b00000010` left to right,
+and bits within a byte right to left).
+
+We proceed through the stream, from which `[0b00010000, 0b00000001]` remains,
+and repeat the operation: the ULEB128-decoding yields `h1 = 16i32`. `16 & 1 == 0`
+and thus the next run is RLE-encoded. We compute the number of repetitions
+via `repetitions = h1 >> 1 = 8`. The body size is `body_length = ceil(bit_width, 8) = 1`,
+which is consistent with `length = 5 = 1 + 2 + 1 + 1`. Since the body is
+`0b00000001 = 1`, we conclude that this run is the number 1 repeated 8 times,
```

Comment:
Is it? `h1 >> 1` is the number of repetitions divided by 8, according to the grammar.

Reply:
Isn't only the bitpacked that is a multiple of 8?

```
rle-run := <rle-header> <repeated-value>
rle-header := varint-encode( (rle-run-len) << 1)
rle-run-len := *see 3 below*
```